lighthouse-core/gather/gatherers/full-page-screenshot.js

```diff
@@ -0,0 +1,132 @@
+   */
+  async _takeScreenshot(passContext, maxScreenshotHeight) {
+    const driver = passContext.driver;
+    const metrics = await driver.sendCommand('Page.getLayoutMetrics');
+    const width = await driver.evaluateAsync('window.innerWidth');
+    const deviceScaleFactor = await driver.evaluateAsync('window.devicePixelRatio');
+    const height = Math.min(metrics.contentSize.height, maxScreenshotHeight);
+
+    await driver.sendCommand('Emulation.setDeviceMetricsOverride', {
+      mobile: passContext.baseArtifacts.TestedAsMobileDevice,
+      height,
+      screenHeight: height,
+      width,
+      screenWidth: width,
+      deviceScaleFactor,
+    });
+
+    const result = await driver.sendCommand('Page.captureScreenshot', {
+      format: 'jpeg',
+      quality: FULL_PAGE_SCREENSHOT_QUALITY,
+    });
+    const data = 'data:image/jpeg;base64,' + result.data;
+
+    return {
+      data,
+      width,
+      height,
+    };
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<LH.Artifacts.FullPageScreenshot | null>}
+   */
+  async afterPass_(passContext) {
+    let screenshot = await this._takeScreenshot(passContext, MAX_SCREENSHOT_HEIGHT);
+
+    if (screenshot.data.length > MAX_DATA_URL_SIZE) {
+      // Hitting the data URL size limit is rare, it only happens for pages on tall
+      // desktop sites with lots of images.
+      // So just cutting down the height a bit usually fixes the issue.
+      screenshot = await this._takeScreenshot(passContext, 5000);
+      if (screenshot.data.length > MAX_DATA_URL_SIZE) {
+        passContext.LighthouseRunWarnings.push('Full page screenshot is too big.');
+        return null;
+      }
+    }
+
+    return screenshot;
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<LH.Artifacts['FullPageScreenshot']>}
+   */
+  async afterPass(passContext) {
+    const {driver} = passContext;
+
+    // In case some other program is controlling emulation, try to remember what the device looks
+    // like now and reset after gatherer is done.
+    const lighthouseControlsEmulation = passContext.settings.emulatedFormFactor !== 'none' &&
+      !passContext.settings.internalDisableDeviceScreenEmulation;
+
+    try {
+      return await this.afterPass_(passContext);
+    } finally {
+      // Revert resized page.
+      if (lighthouseControlsEmulation) {
+        await driver.beginEmulation(passContext.settings);
+      } else {
+        // Best effort to reset emulation to what it was.
+        // https://github.com/GoogleChrome/lighthouse/pull/10716#discussion_r428970681
+        // TODO: seems like this would be brittle. Should at least work for devtools, but what
+        // about scripted puppeteer usages? Better to introduce a "setEmulation" callback
+        // in the LH runner api, which for ex. puppeteer consumers would setup puppeteer emulation,
+        // and then just call that to reset?
```

Comment:
seems like this is another case for the lighthouse signal to go up for major users in addition to the LHR size thing? Anyone setting custom emulation via the protocol could potentially be tripped up (or could demonstrate that this is sufficient)